File: go/0530E-二叉搜索树的最小绝对差.go
/*
 * @Author: SukiEva
 * @Date: 2021-11-11 09:36:40
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */

package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func getMinimumDifference(root *TreeNode) int {
	var pre *TreeNode
	ans := int(1e5)
	var inOrder func(node *TreeNode)
	inOrder = func(node *TreeNode) {
		if node == nil {
			return
		}
		inOrder(node.Left)
		if pre != nil {
			if node.Val-pre.Val < ans {
				ans = node.Val - pre.Val
			}
		}
		pre = node
		inOrder(node.Right)
	}
	inOrder(root)
	return ans
}

File: cpp/0530E-二叉搜索树的最小绝对差.cpp
/*
 * @Author: SukiEva
 * @Date: 2021-12-22 10:35:13
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class Solution {
   public:
    int pre = -1;
    int ans = 1e5 + 10;

    void traveral(TreeNode* root) {
        if (root == nullptr) return;
        traveral(root->left);
        if (pre != -1) ans = min(ans, root->val - pre);
        pre = root->val;
        traveral(root->right);
    }

    int getMinimumDifference(TreeNode* root) {
        traveral(root);
        return ans;
    }
};